// @latticexyz/schema-type/dist/typescript/SchemaType.js
// WARNING: SchemaType enum MUST mirror the solidity version!
// WARNING: SchemaType methods use hardcoded enum indexes, review them after any changes to the enum
export var SchemaType;
(function (SchemaType) {
    SchemaType[SchemaType["UINT8"] = 0] = "UINT8";
    SchemaType[SchemaType["UINT16"] = 1] = "UINT16";
    SchemaType[SchemaType["UINT24"] = 2] = "UINT24";
    SchemaType[SchemaType["UINT32"] = 3] = "UINT32";
    SchemaType[SchemaType["UINT40"] = 4] = "UINT40";
    SchemaType[SchemaType["UINT48"] = 5] = "UINT48";
    SchemaType[SchemaType["UINT56"] = 6] = "UINT56";
    SchemaType[SchemaType["UINT64"] = 7] = "UINT64";
    SchemaType[SchemaType["UINT72"] = 8] = "UINT72";
    SchemaType[SchemaType["UINT80"] = 9] = "UINT80";
    SchemaType[SchemaType["UINT88"] = 10] = "UINT88";
    SchemaType[SchemaType["UINT96"] = 11] = "UINT96";
    SchemaType[SchemaType["UINT104"] = 12] = "UINT104";
    SchemaType[SchemaType["UINT112"] = 13] = "UINT112";
    SchemaType[SchemaType["UINT120"] = 14] = "UINT120";
    SchemaType[SchemaType["UINT128"] = 15] = "UINT128";
    SchemaType[SchemaType["UINT136"] = 16] = "UINT136";
    SchemaType[SchemaType["UINT144"] = 17] = "UINT144";
    SchemaType[SchemaType["UINT152"] = 18] = "UINT152";
    SchemaType[SchemaType["UINT160"] = 19] = "UINT160";
    SchemaType[SchemaType["UINT168"] = 20] = "UINT168";
    SchemaType[SchemaType["UINT176"] = 21] = "UINT176";
    SchemaType[SchemaType["UINT184"] = 22] = "UINT184";
    SchemaType[SchemaType["UINT192"] = 23] = "UINT192";
    SchemaType[SchemaType["UINT200"] = 24] = "UINT200";
    SchemaType[SchemaType["UINT208"] = 25] = "UINT208";
    SchemaType[SchemaType["UINT216"] = 26] = "UINT216";
    SchemaType[SchemaType["UINT224"] = 27] = "UINT224";
    SchemaType[SchemaType["UINT232"] = 28] = "UINT232";
    SchemaType[SchemaType["UINT240"] = 29] = "UINT240";
    SchemaType[SchemaType["UINT248"] = 30] = "UINT248";
    SchemaType[SchemaType["UINT256"] = 31] = "UINT256";
    SchemaType[SchemaType["INT8"] = 32] = "INT8";
    SchemaType[SchemaType["INT16"] = 33] = "INT16";
    SchemaType[SchemaType["INT24"] = 34] = "INT24";
    SchemaType[SchemaType["INT32"] = 35] = "INT32";
    SchemaType[SchemaType["INT40"] = 36] = "INT40";
    SchemaType[SchemaType["INT48"] = 37] = "INT48";
    SchemaType[SchemaType["INT56"] = 38] = "INT56";
    SchemaType[SchemaType["INT64"] = 39] = "INT64";
    SchemaType[SchemaType["INT72"] = 40] = "INT72";
    SchemaType[SchemaType["INT80"] = 41] = "INT80";
    SchemaType[SchemaType["INT88"] = 42] = "INT88";
    SchemaType[SchemaType["INT96"] = 43] = "INT96";
    SchemaType[SchemaType["INT104"] = 44] = "INT104";
    SchemaType[SchemaType["INT112"] = 45] = "INT112";
    SchemaType[SchemaType["INT120"] = 46] = "INT120";
    SchemaType[SchemaType["INT128"] = 47] = "INT128";
    SchemaType[SchemaType["INT136"] = 48] = "INT136";
    SchemaType[SchemaType["INT144"] = 49] = "INT144";
    SchemaType[SchemaType["INT152"] = 50] = "INT152";
    SchemaType[SchemaType["INT160"] = 51] = "INT160";
    SchemaType[SchemaType["INT168"] = 52] = "INT168";
    SchemaType[SchemaType["INT176"] = 53] = "INT176";
    SchemaType[SchemaType["INT184"] = 54] = "INT184";
    SchemaType[SchemaType["INT192"] = 55] = "INT192";
    SchemaType[SchemaType["INT200"] = 56] = "INT200";
    SchemaType[SchemaType["INT208"] = 57] = "INT208";
    SchemaType[SchemaType["INT216"] = 58] = "INT216";
    SchemaType[SchemaType["INT224"] = 59] = "INT224";
    SchemaType[SchemaType["INT232"] = 60] = "INT232";
    SchemaType[SchemaType["INT240"] = 61] = "INT240";
    SchemaType[SchemaType["INT248"] = 62] = "INT248";
    SchemaType[SchemaType["INT256"] = 63] = "INT256";
    SchemaType[SchemaType["BYTES1"] = 64] = "BYTES1";
    SchemaType[SchemaType["BYTES2"] = 65] = "BYTES2";
    SchemaType[SchemaType["BYTES3"] = 66] = "BYTES3";
    SchemaType[SchemaType["BYTES4"] = 67] = "BYTES4";
    SchemaType[SchemaType["BYTES5"] = 68] = "BYTES5";
    SchemaType[SchemaType["BYTES6"] = 69] = "BYTES6";
    SchemaType[SchemaType["BYTES7"] = 70] = "BYTES7";
    SchemaType[SchemaType["BYTES8"] = 71] = "BYTES8";
    SchemaType[SchemaType["BYTES9"] = 72] = "BYTES9";
    SchemaType[SchemaType["BYTES10"] = 73] = "BYTES10";
    SchemaType[SchemaType["BYTES11"] = 74] = "BYTES11";
    SchemaType[SchemaType["BYTES12"] = 75] = "BYTES12";
    SchemaType[SchemaType["BYTES13"] = 76] = "BYTES13";
    SchemaType[SchemaType["BYTES14"] = 77] = "BYTES14";
    SchemaType[SchemaType["BYTES15"] = 78] = "BYTES15";
    SchemaType[SchemaType["BYTES16"] = 79] = "BYTES16";
    SchemaType[SchemaType["BYTES17"] = 80] = "BYTES17";
    SchemaType[SchemaType["BYTES18"] = 81] = "BYTES18";
    SchemaType[SchemaType["BYTES19"] = 82] = "BYTES19";
    SchemaType[SchemaType["BYTES20"] = 83] = "BYTES20";
    SchemaType[SchemaType["BYTES21"] = 84] = "BYTES21";
    SchemaType[SchemaType["BYTES22"] = 85] = "BYTES22";
    SchemaType[SchemaType["BYTES23"] = 86] = "BYTES23";
    SchemaType[SchemaType["BYTES24"] = 87] = "BYTES24";
    SchemaType[SchemaType["BYTES25"] = 88] = "BYTES25";
    SchemaType[SchemaType["BYTES26"] = 89] = "BYTES26";
    SchemaType[SchemaType["BYTES27"] = 90] = "BYTES27";
    SchemaType[SchemaType["BYTES28"] = 91] = "BYTES28";
    SchemaType[SchemaType["BYTES29"] = 92] = "BYTES29";
    SchemaType[SchemaType["BYTES30"] = 93] = "BYTES30";
    SchemaType[SchemaType["BYTES31"] = 94] = "BYTES31";
    SchemaType[SchemaType["BYTES32"] = 95] = "BYTES32";
    SchemaType[SchemaType["BOOL"] = 96] = "BOOL";
    SchemaType[SchemaType["ADDRESS"] = 97] = "ADDRESS";
    SchemaType[SchemaType["UINT8_ARRAY"] = 98] = "UINT8_ARRAY";
    SchemaType[SchemaType["UINT16_ARRAY"] = 99] = "UINT16_ARRAY";
    SchemaType[SchemaType["UINT24_ARRAY"] = 100] = "UINT24_ARRAY";
    SchemaType[SchemaType["UINT32_ARRAY"] = 101] = "UINT32_ARRAY";
    SchemaType[SchemaType["UINT40_ARRAY"] = 102] = "UINT40_ARRAY";
    SchemaType[SchemaType["UINT48_ARRAY"] = 103] = "UINT48_ARRAY";
    SchemaType[SchemaType["UINT56_ARRAY"] = 104] = "UINT56_ARRAY";
    SchemaType[SchemaType["UINT64_ARRAY"] = 105] = "UINT64_ARRAY";
    SchemaType[SchemaType["UINT72_ARRAY"] = 106] = "UINT72_ARRAY";
    SchemaType[SchemaType["UINT80_ARRAY"] = 107] = "UINT80_ARRAY";
    SchemaType[SchemaType["UINT88_ARRAY"] = 108] = "UINT88_ARRAY";
    SchemaType[SchemaType["UINT96_ARRAY"] = 109] = "UINT96_ARRAY";
    SchemaType[SchemaType["UINT104_ARRAY"] = 110] = "UINT104_ARRAY";
    SchemaType[SchemaType["UINT112_ARRAY"] = 111] = "UINT112_ARRAY";
    SchemaType[SchemaType["UINT120_ARRAY"] = 112] = "UINT120_ARRAY";
    SchemaType[SchemaType["UINT128_ARRAY"] = 113] = "UINT128_ARRAY";
    SchemaType[SchemaType["UINT136_ARRAY"] = 114] = "UINT136_ARRAY";
    SchemaType[SchemaType["UINT144_ARRAY"] = 115] = "UINT144_ARRAY";
    SchemaType[SchemaType["UINT152_ARRAY"] = 116] = "UINT152_ARRAY";
    SchemaType[SchemaType["UINT160_ARRAY"] = 117] = "UINT160_ARRAY";
    SchemaType[SchemaType["UINT168_ARRAY"] = 118] = "UINT168_ARRAY";
    SchemaType[SchemaType["UINT176_ARRAY"] = 119] = "UINT176_ARRAY";
    SchemaType[SchemaType["UINT184_ARRAY"] = 120] = "UINT184_ARRAY";
    SchemaType[SchemaType["UINT192_ARRAY"] = 121] = "UINT192_ARRAY";
    SchemaType[SchemaType["UINT200_ARRAY"] = 122] = "UINT200_ARRAY";
    SchemaType[SchemaType["UINT208_ARRAY"] = 123] = "UINT208_ARRAY";
    SchemaType[SchemaType["UINT216_ARRAY"] = 124] = "UINT216_ARRAY";
    SchemaType[SchemaType["UINT224_ARRAY"] = 125] = "UINT224_ARRAY";
    SchemaType[SchemaType["UINT232_ARRAY"] = 126] = "UINT232_ARRAY";
    SchemaType[SchemaType["UINT240_ARRAY"] = 127] = "UINT240_ARRAY";
    SchemaType[SchemaType["UINT248_ARRAY"] = 128] = "UINT248_ARRAY";
    SchemaType[SchemaType["UINT256_ARRAY"] = 129] = "UINT256_ARRAY";
    SchemaType[SchemaType["INT8_ARRAY"] = 130] = "INT8_ARRAY";
    SchemaType[SchemaType["INT16_ARRAY"] = 131] = "INT16_ARRAY";
    SchemaType[SchemaType["INT24_ARRAY"] = 132] = "INT24_ARRAY";
    SchemaType[SchemaType["INT32_ARRAY"] = 133] = "INT32_ARRAY";
    SchemaType[SchemaType["INT40_ARRAY"] = 134] = "INT40_ARRAY";
    SchemaType[SchemaType["INT48_ARRAY"] = 135] = "INT48_ARRAY";
    SchemaType[SchemaType["INT56_ARRAY"] = 136] = "INT56_ARRAY";
    SchemaType[SchemaType["INT64_ARRAY"] = 137] = "INT64_ARRAY";
    SchemaType[SchemaType["INT72_ARRAY"] = 138] = "INT72_ARRAY";
    SchemaType[SchemaType["INT80_ARRAY"] = 139] = "INT80_ARRAY";
    SchemaType[SchemaType["INT88_ARRAY"] = 140] = "INT88_ARRAY";
    SchemaType[SchemaType["INT96_ARRAY"] = 141] = "INT96_ARRAY";
    SchemaType[SchemaType["INT104_ARRAY"] = 142] = "INT104_ARRAY";
    SchemaType[SchemaType["INT112_ARRAY"] = 143] = "INT112_ARRAY";
    SchemaType[SchemaType["INT120_ARRAY"] = 144] = "INT120_ARRAY";
    SchemaType[SchemaType["INT128_ARRAY"] = 145] = "INT128_ARRAY";
    SchemaType[SchemaType["INT136_ARRAY"] = 146] = "INT136_ARRAY";
    SchemaType[SchemaType["INT144_ARRAY"] = 147] = "INT144_ARRAY";
    SchemaType[SchemaType["INT152_ARRAY"] = 148] = "INT152_ARRAY";
    SchemaType[SchemaType["INT160_ARRAY"] = 149] = "INT160_ARRAY";
    SchemaType[SchemaType["INT168_ARRAY"] = 150] = "INT168_ARRAY";
    SchemaType[SchemaType["INT176_ARRAY"] = 151] = "INT176_ARRAY";
    SchemaType[SchemaType["INT184_ARRAY"] = 152] = "INT184_ARRAY";
    SchemaType[SchemaType["INT192_ARRAY"] = 153] = "INT192_ARRAY";
    SchemaType[SchemaType["INT200_ARRAY"] = 154] = "INT200_ARRAY";
    SchemaType[SchemaType["INT208_ARRAY"] = 155] = "INT208_ARRAY";
    SchemaType[SchemaType["INT216_ARRAY"] = 156] = "INT216_ARRAY";
    SchemaType[SchemaType["INT224_ARRAY"] = 157] = "INT224_ARRAY";
    SchemaType[SchemaType["INT232_ARRAY"] = 158] = "INT232_ARRAY";
    SchemaType[SchemaType["INT240_ARRAY"] = 159] = "INT240_ARRAY";
    SchemaType[SchemaType["INT248_ARRAY"] = 160] = "INT248_ARRAY";
    SchemaType[SchemaType["INT256_ARRAY"] = 161] = "INT256_ARRAY";
    SchemaType[SchemaType["BYTES1_ARRAY"] = 162] = "BYTES1_ARRAY";
    SchemaType[SchemaType["BYTES2_ARRAY"] = 163] = "BYTES2_ARRAY";
    SchemaType[SchemaType["BYTES3_ARRAY"] = 164] = "BYTES3_ARRAY";
    SchemaType[SchemaType["BYTES4_ARRAY"] = 165] = "BYTES4_ARRAY";
    SchemaType[SchemaType["BYTES5_ARRAY"] = 166] = "BYTES5_ARRAY";
    SchemaType[SchemaType["BYTES6_ARRAY"] = 167] = "BYTES6_ARRAY";
    SchemaType[SchemaType["BYTES7_ARRAY"] = 168] = "BYTES7_ARRAY";
    SchemaType[SchemaType["BYTES8_ARRAY"] = 169] = "BYTES8_ARRAY";
    SchemaType[SchemaType["BYTES9_ARRAY"] = 170] = "BYTES9_ARRAY";
    SchemaType[SchemaType["BYTES10_ARRAY"] = 171] = "BYTES10_ARRAY";
    SchemaType[SchemaType["BYTES11_ARRAY"] = 172] = "BYTES11_ARRAY";
    SchemaType[SchemaType["BYTES12_ARRAY"] = 173] = "BYTES12_ARRAY";
    SchemaType[SchemaType["BYTES13_ARRAY"] = 174] = "BYTES13_ARRAY";
    SchemaType[SchemaType["BYTES14_ARRAY"] = 175] = "BYTES14_ARRAY";
    SchemaType[SchemaType["BYTES15_ARRAY"] = 176] = "BYTES15_ARRAY";
    SchemaType[SchemaType["BYTES16_ARRAY"] = 177] = "BYTES16_ARRAY";
    SchemaType[SchemaType["BYTES17_ARRAY"] = 178] = "BYTES17_ARRAY";
    SchemaType[SchemaType["BYTES18_ARRAY"] = 179] = "BYTES18_ARRAY";
    SchemaType[SchemaType["BYTES19_ARRAY"] = 180] = "BYTES19_ARRAY";
    SchemaType[SchemaType["BYTES20_ARRAY"] = 181] = "BYTES20_ARRAY";
    SchemaType[SchemaType["BYTES21_ARRAY"] = 182] = "BYTES21_ARRAY";
    SchemaType[SchemaType["BYTES22_ARRAY"] = 183] = "BYTES22_ARRAY";
    SchemaType[SchemaType["BYTES23_ARRAY"] = 184] = "BYTES23_ARRAY";
    SchemaType[SchemaType["BYTES24_ARRAY"] = 185] = "BYTES24_ARRAY";
    SchemaType[SchemaType["BYTES25_ARRAY"] = 186] = "BYTES25_ARRAY";
    SchemaType[SchemaType["BYTES26_ARRAY"] = 187] = "BYTES26_ARRAY";
    SchemaType[SchemaType["BYTES27_ARRAY"] = 188] = "BYTES27_ARRAY";
    SchemaType[SchemaType["BYTES28_ARRAY"] = 189] = "BYTES28_ARRAY";
    SchemaType[SchemaType["BYTES29_ARRAY"] = 190] = "BYTES29_ARRAY";
    SchemaType[SchemaType["BYTES30_ARRAY"] = 191] = "BYTES30_ARRAY";
    SchemaType[SchemaType["BYTES31_ARRAY"] = 192] = "BYTES31_ARRAY";
    SchemaType[SchemaType["BYTES32_ARRAY"] = 193] = "BYTES32_ARRAY";
    SchemaType[SchemaType["BOOL_ARRAY"] = 194] = "BOOL_ARRAY";
    SchemaType[SchemaType["ADDRESS_ARRAY"] = 195] = "ADDRESS_ARRAY";
    SchemaType[SchemaType["BYTES"] = 196] = "BYTES";
    SchemaType[SchemaType["STRING"] = 197] = "STRING";
})(SchemaType || (SchemaType = {}));
